units/event_metajob.go

```diff
@@ -159,28 +162,51 @@ func (j *eventMetaJob) dispatchLoop(ctx context.Context) error {
 	startTime := time.Now()
 	logger := event.NewDBEventLogger(event.AllLogCollection)
 	catcher := grip.NewSimpleCatcher()
-	notifications := make([][]notification.Notification, len(j.events))
 
-	var err error
-	for i := range j.events {
-		notifications[i], err = tryProcessOneEvent(&j.events[i])
-		catcher.Add(err)
-		if err = notification.InsertMany(notifications[i]...); err != nil {
-			grip.Error(message.WrapError(err, message.Fields{
-				"job_id":        j.ID(),
-				"job":           eventMetaJobName,
-				"source":        "events-processing",
-				"notifications": notifications[i],
-				"message":       "can't insert notifications",
-			}))
-			catcher.Add(err)
+	input := make(chan event.EventLogEntry, len(j.events))
```

Comment:
I don't follow: why do we write everything into the channel before starting the goroutines to process?

Reply:
because reads from channels are thread safe and slice access aren't.

Reply:
I understand why we need to copy into a channel and process out of the channel, but why don't we do it after starting to read out of the channel so we can process as we write into the channel?

Reply:
I think we'd need to have an additional thread in the case that you note.

Reply:
I was envisioning the main thread doing the writing and each goroutine pulling off the channel.

Reply:
That's what this is doing. I think of the input channel as a thread-safe slice here

Reply:
The consumer threads should work as the events are being produced.
It doesn't really matter: copying into the channel shouldn't take long.